CMSIS-NN conv, depthwise conv kernels: allocate per channel memory dynamically (#42748)

janjongboom opened the pull request:
See #42748 

The CMSIS-NN convolutional and depthwise convolutional kernels write in uninitialized memory when num_channels is higher than 256. This is because the OpData struct allocates memory statically, rather than request memory from the arena, like the non-CMSIS-NN kernels. As there is no boundary checking in this function this leads to errors when dealing with any network that contains more channels in a layer (this has shown for us in e.g. MobileNet v2 0.05). This patch brings the behavior of the kernel in line with the normal TFLM conv and depthwise_conv kernels, by allocating the memory from the arena.

Additionally I've (in the second commit) changed the way that kernels allocate memory from using the scratch buffer to allocating persistent memory. As far as I can see this should be the case anyway, this memory is required permanently for the kernel.

I've [backported](https://github.com/edgeimpulse/tensorflow/tree/fix-42748-v2.3.0) this fix to the TensorFlow v2.3.0 release and verified that it works on Mbed OS 6.2 on the ST IoT Discovery Kit using CMSIS-NN (compiled with GCC ARM 9), but current master hardfaults on a nullptr reference both with and without CMSIS-NN.

/cc @kwagyeman @dansitu

janjongboom commented:
@googlebot I signed it!

felix-johnny commented:
@janjongboom  Thanks for this. I had to revert the dynamic allocation due to a bug recently for DW conv.  I am adding a reference of that PR here for context just to be sure that that the bug is resolved https://github.com/tensorflow/tensorflow/pull/41121 .

 I see that in the bug report you mention that there are no boundary checks.. There is a boundary check atleast for DW conv. Did that not fault?

mansnils commented:
@janjongboom  Thanks for this. The RequestScratchBufferInArena/GetScratchBuffer should not be removed. It should still be used for CMSIS scratch buffer. Whenever Scratch buffer API can be used it should be since it has not the same life time as the interpreter, like AllocatePersistentBuffer and hence memory can be reused. The CMSIS API expects a scratch buffer and then RequestScratchBufferInArena/GetScratchBuffer is perfect.
Until recently it was not possibly to have both AllocatePersistentBuffer and RequestScratchBufferInArena mixed but now it should be possible since this PR is merged: https://github.com/tensorflow/tensorflow/commit/59d177d9acabe8e70bc33e554a364d2620bc6999
So please add back the RequestScratchBufferInArena/GetScratchBuffer. Thanks.

janjongboom commented:
@masnils, interesting. Yes, I ran into the issue of not being able to use both with the micro allocator in TF 2.3.0 (our custom allocator does not have this issue) and backported our changes in that branch to this PR. Will revert here.

janjongboom committed:
CMSIS-NN conv, depthwise conv kernels: allocate per channel memory dynamically (fix #42748)

The CMSIS-NN convolutional and depthwise convolutional kernels write in uninitialized memory when num_channels is higher than 256. This is because the OpData struct allocates memory statically, rather than request memory from the arena, like the non-CMSIS-NN kernels. As there is no boundary checking in this function this leads to errors when dealing with any network that contains more channels in a layer (this has shown for us in e.g. MobileNet v2 0.05). This patch brings the behavior of the kernel in line with the normal TFLM conv and depthwise_conv kernels, by allocating the memory from the arena.

This reverts commit 5986dd0525094a8295a9feca4ad65023a56bcc8e.

mansnils commented on tensorflow/lite/micro/kernels/cmsis-nn/depthwise_conv.cc:
```diff
@@ -134,7 +133,6 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
     // Allocate memory for per-channel quantization parameters
     const int num_channels =
```
num_channels here is not used and should be removed.

mansnils replied:
Sorry, it's the other way around, the allocation should be moved here under (input->type == kTfLiteInt8). So the other num_channels below can be removed.

mansnils reviewed (commented):
Great. Thanks for the update, I added a few comments

felix-johnny commented on tensorflow/lite/micro/kernels/cmsis-nn/conv.cc:
```diff
@@ -159,6 +157,18 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
   output_dims.w = output->dims->data[2];
   output_dims.c = output_shape.Dims(3);
 
+  // Dynamically allocate per-channel quantization parameters.
```
Please move the whole block inside the  if (input->type == kTfLiteInt8) statement in line 180 as we don't want to allocate this in the case of a uint8 model. This piece of code is common for both int8 and uint8 case.

janjongboom commented:
@mansnils @felix-johnny See my new commit. `CalculateOpData` does pass in these parameters so I moved the if-clause before that call. I'm wondering though whether it's really only applicable for `i8` models, as these parameters are referenced in depthwise_conv.cc for all models that are not f32:

```
  // Note that quantized inference requires that all tensors have their
  // parameters set. This is usually done during quantized training.
  if (data_type != kTfLiteFloat32) {
    const TfLiteTensor* input = GetInput(context, node, kInputTensor);
    const TfLiteTensor* filter = GetInput(context, node, kFilterTensor);
    const TfLiteTensor* bias =
        GetOptionalInputTensor(context, node, kBiasTensor);
    TfLiteTensor* output = GetOutput(context, node, kOutputTensor);
    int num_channels = filter->dims->data[kDepthwiseConvQuantizedDimension];

    return tflite::PopulateConvolutionQuantizationParams(
        context, input, filter, bias, output, params->activation,
        &data->output_multiplier, &data->output_shift,
        &data->output_activation_min, &data->output_activation_max,
        data->per_channel_output_multiplier,
        reinterpret_cast<int*>(data->per_channel_output_shift), num_channels);
  }
```

felix-johnny commented:
@janjongboom Unfortunately for now, the allocation has to be moved out of the if(int8) (sorry, despite saying the opposite earlier) as the per channel(int<>) only params are accessed even in the per-tensor cases. This leads to a crash for uint8 models. While this results in an over allocation of memory for non-int<8,16> models, it is an acceptable temporary option 
compared to a crash. I filed a ticket about that today.  https://github.com/tensorflow/tensorflow/issues/42883

kwagyeman commented:
@janjongboom Can you update the code to get this merged? Or, what branch should OpenMV follow? We want to integrate CMSIS-NN now.

janjongboom commented:
@kwagyeman Yes I'll update this PR. Our release branch is https://github.com/edgeimpulse/tensorflow/tree/fix-42748-v2.3.0, this has been verified on all typical Edge Impulse models on Cortex-M4F and Cortex-M7 physical hardware.

janjongboom committed:
Allocate per_channel_output parameters only for i8 models

janjongboom commented:
@felix-johnny Done!

advaitjain reviewed (changes requested):
Let's wait for a bit before merging this PR. The kernel_conv_test for CMSIS-NN is currently failing and is disabled with PR #43163 

I'd like to have that problem fixed first to avoid having multiple regressions at the same time.

@janjongboom: if you're up for it, you can make a separate PR for the depthwise_conv change and that can be merged without delay since the depthwise_conv test is passing.

Lastly, I would recommend adding to the PR description something to confirm that the tests were run. Due to the limitation of our current CI system the cmsis-nn tests do not get run as part of the presubmit checks (they are only built).

Which in this case would be:

Manually ran the following command and confirmed that the test passes:
```
make -f tensorflow/lite/micro/kernels/tools/make/Makefile TARGET=stm32f4 TAGS=cmsis-nn test_kernel_depthwise_conv_test
```

kwagyeman commented:
All, separate topic. But, can you shrink the int8 person detector model to be the same size as the uint8 person detector model and replace the original? I can't fit the larger int8 model in our flash. But, it runs at 12 FPS on the STM32H7 versus 6 FPS of the uint8 model.

advaitjain commented:
> All, separate topic. But, can someone shrink the int8 person detector model to be the same size as the uint8 person detector model and replace the original? I can't fit the larger int8 model in our flash. But, it runs at 12 FPS on the STM32H7 versus 6 FPS of the uint8 model.

If you could make a github issue for this that would be great. Thanks!

felix-johnny commented:
@advaitjain The failing case for conv is because the dilation case isn't handled by the cmsis-nn/conv.cc glue like it is done in the dw conv case. The conv.cc glue is long due for an update. My apologies for that.

@janjongboom  If you can make this PR just about DW conv like Advait suggested, I can do the overhaul of the conv.cc along with your current change and also fix https://github.com/tensorflow/tensorflow/issues/42951 and the failing unit test. I can make the PR in a couple of days. That ok?

advaitjain commented:
> @advaitjain The failing case for conv is because the dilation case isn't handled by the cmsis-nn/conv.cc glue like it is done in the dw conv case. The conv.cc glue is long due for an update. My apologies for that.
> 

That would be excellent!

> @janjongboom If you can make this PR just about DW conv like Advait suggested, I can do the overhaul of the conv.cc along with your current change and also fix #42951 and the failing unit test. I can make the PR in a couple of days. That ok?

This sounds good too. Though I would strongly request separating these into [smaller PRs](https://google.github.io/eng-practices/review/developer/small-cls.html) that:
  1. Fix the dilation test case
  1. Fix #42951
  1. Implement the change from the current PR.
 
And the moment we have the conv test passing, we should remove the exclusion from here:
https://github.com/tensorflow/tensorflow/blob/1a3fcd65dbfdd1439f8834725a34d4e11a437565/tensorflow/lite/micro/tools/make/targets/stm32f4_makefile.inc#L66-L74

felix-johnny commented:
@advaitjain Sure. I've made 3 PR's

1. M  PR https://github.com/tensorflow/tensorflow/pull/43484 : Fix #42951
2. XS PR https://github.com/tensorflow/tensorflow/pull/43485 : Fix dilation case (Needs first PR to make the fix complete)
3. S   PR https://github.com/tensorflow/tensorflow/pull/43486 : implement change from this PR.

The order in which it goes in isn't important as I realize that I'll have to to two manual rebases irrespective of the order.

janjongboom committed:
Revert changes in conv.cc

janjongboom commented:
@felix-johnny  done!

felix-johnny commented:
Thanks @janjongboom .. The PR for conv is up for review. So, we should be good on both soon!

felix-johnny commented:
@advaitjain Could you please have a look at the updated split?

advaitjain reviewed (approved):
apologies for the delay. looks good!
